src/point.ts

```diff
@@ -1,36 +1,52 @@
-/* @flow -*- mode: flow -*- */
+import { eqNum } from "./util/number"
 
-type Point = {
+export interface Point {
   x: number;
   y: number;
 };
 
+
 export function point(x: number, y: number): Point {
-  return { x: x, y: y };
+  return { x, y };
+}
+
+// DuckTyped
```

Comment:
can you explain whats going on here? p is any, but also Point?

Reply:
So this is what's known as "any type", which you can read about [here](https://www.typescriptlang.org/docs/handbook/advanced-types.html#typeof-type-guards)

The short of it is, this is what's known of as a type predicate, it checks if a var (which could be `any`) type is another type. So if you look at the return type it's `type is Point`, which basically means that this function will return a boolean that indicates whether arg p is a `Point` :)

Reply:
I....still don't get it, even after reading that example they give....whats the difference between  specifying the return as "p is Point" and just specifying "Boolean"?

Reply:
Alright, so the coolest part comes afterwards! Let's do some... Coding!

So imagine you have a function that takes an animal, and makes it... Do it's thing
```typescript
function doCall() {}
```

Reply:
Okay! Let's do some... Coding!

Imagine you have a function, that takes as its argument an animal, and it makes the animal sound its sounds:

```typescript
class Animal {
  doCall() {}
}

function doCall(animal: Animal) {
  console.log(animal.doCall())
}
```

But more realistically, as we are all taught in our schools, different animals make different calls (Yes I am aware this is bad OOP): 

```typescript
class Dog {
  bark() {
    return "aroof"
  }
}


class Dolphin {
  eek() {
    return "eek eek prakk"
  }
}
```

So just calling `doCall()` and calling it a day might not be good enough, so we union the different types of animals here

```typescript
function doCall(animal: Dolphin | Dog) {
  if(animal.eek) {
    animal.eek()
  } else if (animal.bark) {
    animal.bark()
  }
}
```

This would however, um, throw an error, it would say:

```
Property 'eek' does not exist on type 'Dog | Dolphin'.
> Property 'eek' does not exist on type 'Dog'.
```

So how do we hint to Typescript what type we're dealing with? Introducing _type predicates_, or Bla is Thing-funcs:

```typescript


function isDolphin(animal: any): animal is Dolphin {
  return animal instanceof Dolphin
}

function isDog(animal: any): animal is Dog {
  return animal instanceof Dog
}

function doCall(animal: Dolphin | Dog) {
  if(isDolphin(animal)) {
    animal.eek()
  } else if(isDog(animal)) {
    animal.bark()
  }
}

```

And that is what those types do, simplified

Reply:
Uh, so there are two _really_ common scenarios where you don't need to type it that way:

one is with `typeof`

```typescript
function concat(a: string | number, b: string | number): string {
  if(typeof a === "string") {
    return (a + b)
  } else if (typeof a === "number") {
    return  a + b // Operator '+' cannot be applied to types 'number'
  }
}
```

Typeof is implemented as a type predicate, so it would do the right thing anyway.


The other is `instanceof`, so the above code would have been (more elegantly solved?) with:

```typescript
function doCall(animal: Dolphin | Dog) {
  if(animal instanceof Dolphin) {
    animal.eek()
  } else if(animal instanceof Dog) {
    animal.bark()
  } 
```

Typescript would have then _known_ that everything within those ifs are their respective instances with their respective properties

Reply:
TLDR; `p is Point` tells typescript specifically that the return is a Point, and that way you can access it's pointy properties. (like `x` and `why`), and it would work everywhere a function expects exactly a point.

Remember that an `Anchored<B>` is defined with `{value: Point | any}`, so knowing if it's a Point or anything else tells us if we can compare `x + y` or need to do some other sort of comparison

Reply:
Haha thanks for that, I didn't think that this
```
function doCall(animal: Dolphin | Dog) {
  if(animal.eek) {
    animal.eek()
  } else if (animal.bark) {
    animal.bark()
  }
}
```
would give you an error with TS.

Reply:
Oh... I guess you're right ;)
<img width="527" alt="Screenshot 2019-09-16 at 16 43 53" src="https://user-images.githubusercontent.com/1070211/64967692-45915f00-d8a1-11e9-972f-affbb266e3ce.png">